### PR TITLE
added SIGLIP + tests - Closes #39

### LINKED ISSUE #39: Add SigLip loss
```import torch
import torch.nn.functional as F

def siglip_loss(image_embeddings, text_embeddings, temperature=1.0):
    """
    Compute the SigLIP loss for a batch of image and text embeddings.

    Args:
        image_embeddings (torch.Tensor): Tensor of shape (batch_size, embedding_dim) containing image embeddings.
        text_embeddings (torch.Tensor): Tensor of shape (batch_size, embedding_dim) containing text embeddings.
        temperature (float): Scaling factor for the logits. Default is 1.0.

    Returns:
        torch.Tensor: Scalar tensor representing the SigLIP loss.
    """
    # Normalize embeddings
    image_embeddings = F.normalize(image_embeddings, p=2, dim=-1)
    text_embeddings = F.normalize(text_embeddings, p=2, dim=-1)

    # Compute pairwise similarity scores
    logits = torch.matmul(image_embeddings, text_embeddings.t()) / temperature

    # Create labels: 1 for positive pairs, 0 for negative pairs
    batch_size = image_embeddings.size(0)
    labels = torch.eye(batch_size, device=logits.device)

    # Compute binary cross-entropy loss with logits
    loss = F.binary_cross_entropy_with_logits(logits, labels)

    return loss```

### PR DESCRIPTION
**Pull request by robertavram-md**
Fixes #39 
This pull request includes significant updates to the configuration files, loss functions, and testing suite to support a new loss function called SIGLIP and its distributed data parallel (DDP) variant. The changes also include updates to the dataset generation and logging utilities.

### Configuration Updates:
* Updated `run_mode` to `train`, changed `data_filename`, and reduced `batch_size` in `config/clip/base_config.yaml`. [[1]](diffhunk://#diff-1cb0c54c851330e53e716cb9bf6757cca3418bf0468c1d2679ebf63a68fdef62L9-R11) [[2]](diffhunk://#diff-1cb0c54c851330e53e716cb9bf6757cca3418bf0468c1d2679ebf63a68fdef62L21-R21)
* Changed `loss_name` to `siglip_ddp` in `config/clip/base_config.yaml`.
* Modified `config/clip/sweep_config_multi_video.yaml` to remove the value for `--nproc_per_node`.

### Loss Function Enhancements:
* Added `SiglipLoss` and `SiglipLossDDP` classes to `utils/loss/losses.py` and registered them with `LossRegistry`. [[1]](diffhunk://#diff-fbac97edf220966c42a7ba300ae27852cc9733a9752e382c8c47c681f546138eR151-R280) [[2]](diffhunk://#diff-fbac97edf220966c42a7ba300ae27852cc9733a9752e382c8c47c681f546138eR303-R308)
* Updated `InfoNCELoss` to support `siglip` loss type.
* Added `SIGLIP` and `SIGLIP_DDP` to `LossType` enum in `utils/enums.py`.

### Testing Suite Additions:
* Introduced comprehensive tests for the new SIGLIP loss functions and their DDP variants in `tests/test_losses.py`.

### Dataset and Logging Updates:
* Updated kernel specification and Python version in `generate_dataset.ipynb`. [[1]](diffhunk://#diff-e57f74d05153cab1d4eaa70d22b51b76d98ec7e3dfe9b18fab35045712953c0aL1531-R1531) [[2]](diffhunk://#diff-e57f74d05153cab1d4eaa70d22b51b76d98ec7e3dfe9b18fab35045712953c0aL1545-R1545)
* Enhanced logger creation to include new configuration parameters in `utils/logging.py`.
* Ensured `loss_name` is included in the configuration dictionary in `utils/wandb_wrapper.py`.

## Summary by Sourcery

Add SIGLIP and SIGLIP_DDP loss functions with corresponding tests and update configuration and logging utilities to support these new loss types.

New Features:
- Introduce SIGLIP and SIGLIP_DDP loss functions for enhanced contrastive learning with a gating mechanism.

Enhancements:
- Update configuration files to support SIGLIP loss functions, including changes to run mode, data filename, batch size, and loss name.
- Enhance logging utilities to include new configuration parameters related to SIGLIP loss functions.

Tests:
- Add comprehensive tests for SIGLIP and SIGLIP_DDP loss functions, ensuring their correct implementation and behavior.

### CODE REVIEW

**Comment by JacquesDelfrate** [comment]
unit test fails.

**Comment by JacquesDelfrate** [comment]
@robertavram-md Fixed the unittest, not true that gated fonction is monotically increase over [-Inf; +Inf]
assuming gated function of x is g(x) = x * Sigmoid(x)
g(x) is <= x for x > 0
and 
g(x) >= x for x < 0
example si x = 2 alors 2 * ~0.88 = 1.76 -> x > g(x)
mais si x = -2 alors -2 * ~0.1 = -0.2 -> x < g(x)

**Comment by JacquesDelfrate** [comment]
Tensor object are not created with variable requires_grad=True by default, this flag needs to be activated to test gradient operation propagation in pytorch

**Comment by JacquesDelfrate** [comment]
Unit test working
